File: lab4/the-triwizard-merge-sort.py
def mergeSort(arr):
    if len(arr) == 1:
        return arr
    mid = len(arr) // 2
    left = mergeSort(arr[:mid])
    right = mergeSort(arr[mid:])
    return merge(left, right)

def merge(arr_l, arr_r):
    i = 0
    j = 0
    result = []
    while i < len(arr_l) and j < len(arr_r):
        if(arr_l[i] < arr_r[j]):
            result.append(arr_l[i])
            i += 1
        else:
            result.append(arr_r[j])
            j += 1

    while i < len(arr_l):
        result.append(arr_l[i])
        i += 1
    
    while j < len(arr_r):
        result.append(arr_r[j])
        j += 1
    
    return result

cases = int(input())

for _ in range(cases):
    arr = [int(i) for i in input().split(' ')]
    arr = mergeSort(arr)
    for i in arr:
        print(i, end=" ")
    print("")

File: lab4/the-triwizard-merge-sort.cpp
#include <iostream>
#include <vector>
#include <sstream>

using namespace std;

vector<int> merge(vector<int> &left, vector<int> &right)
{
    vector<int> result;
    int i = 0, j = 0;
    while (i < left.size() && j < right.size())
    {
        if (left[i] <= right[j])
        {
            result.push_back(left[i]);
            i++;
        }
        else
        {
            result.push_back(right[j]);
            j++;
        }
    }
    while (i < left.size())
    {
        result.push_back(left[i]);
        i++;
    }
    while (j < right.size())
    {
        result.push_back(right[j]);
        j++;
    }
    return result;
}

vector<int> mergeSort(vector<int> &arr)
{
    if (arr.size() <= 1)
        return arr;
    int mid = arr.size() / 2;
    vector<int> left(arr.begin(), arr.begin() + mid);
    vector<int> right(arr.begin() + mid, arr.end());
    left = mergeSort(left);
    right = mergeSort(right);
    return merge(left, right);
}

int main()
{
    int n;
    cin >> n;
    cin.ignore();
    for (int i = 0; i < n; i++)
    {
        string line;
        getline(cin, line);
        stringstream ss(line);
        vector<int> arr;
        int x;
        while (ss >> x)
        {
            arr.push_back(x);
        }
        arr = mergeSort(arr);
        for (int j = 0; j < arr.size(); j++)
        {
            cout << arr[j] << " ";
        }
        cout << endl;
    }
    return 0;
}